### PR TITLE
Add aggregate function countEqualRanges

### PR DESCRIPTION
**Pull request by hczhcz**
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

Changelog category (leave one):
- New Feature

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):

Add aggregate function `countEqualRanges` to count segments with equal values

Detailed description / Documentation draft:

> # countEqualRanges
>
> Calculates the number of segments with equal values in a column.
>
> ``` sql
> countEqualRanges(x)
> ```

``` sql
SELECT countEqualRanges(x) FROM (
    SELECT arrayJoin(['a', 'a', 'a', 'b', 'c', 'c', 'a', 'd', 'd']) AS x
)
```

``` text
┌─countEqualRanges(x)─┐
│                   5 │
└─────────────────────┘
```

### CODE REVIEW

File: src/AggregateFunctions/AggregateFunctionCountEqualRanges.h [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+    DataTypePtr getReturnType() const override
+    {
+        return std::make_shared<DataTypeUInt64>();
+    }
+
+    void ALWAYS_INLINE add(AggregateDataPtr place, const IColumn ** columns, size_t row_num, Arena *) const override
+    {
+        const Field & value = (*columns[0])[row_num];
+
+        if (!this->data(place).count || this->data(place).value != value)
```
Comment (amosbird) [inline src/AggregateFunctions/AggregateFunctionCountEqualRanges.h:47]: The definition of segment is confusing here. It relies on the order of value delivery.

**Comment by Avogar** [comment]
It won't work correctly in case when we select data without any subquery (directly from table) due to parallel aggregation. Example:
```
:) CREATE TABLE test_equal_ranges (x UInt32) ENGINE=MergeTree() ORDER BY x;

CREATE TABLE test_equal_ranges
(
    `x` UInt32
)
ENGINE = MergeTree()
ORDER BY x

Query id: 84ff59a7-ca72-4d51-b261-d07d6dc43019

Ok.

0 rows in set. Elapsed: 0.067 sec. 

:) INSERT INTO test_equal_ranges SELECT 0 FROM numbers(10000000)

INSERT INTO test_equal_ranges SELECT 0
FROM numbers(10000000)

Query id: 4efe4309-e4fc-439f-8e5c-1c7d26f0f4a9

Ok.

0 rows in set. Elapsed: 0.895 sec. Processed 10.49 million rows, 83.88 MB (11.71 million rows/s., 93.68 MB/s.) 

:) SELECT countEqualRanges(x) FROM test_equal_ranges

SELECT countEqualRanges(x)
FROM test_equal_ranges

Query id: f57e6d54-8668-47d2-a196-0990d11f59b2

┌─countEqualRanges(x)─┐
│                  50 │
└─────────────────────┘

1 rows in set. Elapsed: 0.108 sec. Processed 10.00 million rows, 40.00 MB (92.64 million rows/s., 370.56 MB/s.)
```

It will work only if limit max_threads to 1:
```
:) SET max_threads=1

SET max_threads = 1

Query id: d7b7333d-3f3c-4af5-97f7-606d6043f6dc

Ok.

0 rows in set. Elapsed: 0.002 sec. 

:) SELECT countEqualRanges(x) FROM test_equal_ranges

SELECT countEqualRanges(x)
FROM test_equal_ranges

Query id: a46d75b6-daf5-400f-8ac7-a8645e247a8f

┌─countEqualRanges(x)─┐
│                   1 │
└─────────────────────┘

1 rows in set. Elapsed: 0.498 sec. Processed 10.00 million rows, 40.00 MB (20.08 million rows/s., 80.32 MB/s.)
```

So, current implementation of merge method doesn't allow to merge aggregation states received from different threads correctly.  
Possible solution: we can have a structure (e.g. binary tree) that stores information about intervals we processed (left and right index, left and right value, count of equal ranges in this interval) sorted by left index. In merge method we can merge these structures and join adjacent intervals where possible, so after the last merge we will have only one interval in our structure with the correct answer. If use binary tree the complexity of merge in time is O(threads_count * log(threads_count)). In this method we use more time in merge and more memory (to store our structure), but it will allow to use this function with parallel aggregation.

**Comment by hczhcz** [comment]
@Avogar 

The actual purpose of this function is mentioned in the docs.

> This function can be used as a faster uniqExact on sorted data.

I made some changes so that the result is consistent with `groupArray`.

```sql
SELECT
    arrayReduce('countEqualRanges', groupArray(x)),
    countEqualRanges(x)
FROM test

Query id: 3c85f554-3f85-4f9b-8ebe-8f592d1a6e9e

┌─arrayReduce('countEqualRanges', groupArray(x))─┬─countEqualRanges(x)─┐
│                                            255 │                 255 │
└────────────────────────────────────────────────┴─────────────────────┘
```

**Comment by Avogar** [comment]
@hczhcz We discussed with @alexey-milovidov such functions that depend on the order. We don't want a common ClickHouse user to be able to use such functions without knowing that they depend on the order. So, we decided that it would be better to add a setting like `use_aggregate_functions_depended_on_order` and use countEqualRanges only if this setting is enabled, otherwise we need throw an exception. Default value for this setting should be 0.

**Comment by hczhcz** [comment]
@Avogar Should we apply this setting to `groupArray*` functions?

**Comment by Avogar** [comment]
> @Avogar Should we apply this setting to `groupArray*` functions?

No, we will apply this setting only for new functions.

**Comment by alexey-milovidov** [comment]
I think our codebase is just not ready for this function.
Correct solution will be the following:
1. @KochetovNicolai will add ordering properties on data streams.
2. Extend the interface of `IAggregateFunction` to support specialized implementation on ordered data.
3. If the data stream is ordered by arguments of `uniq*` function, then we can apply optimized version of this function.

**Pull request closed without merging**